---
title: "Vaccine effectiveness"
author: "Rob Challen"
date: "10/06/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = TRUE,
  warning = TRUE
)

# The `here` library sorts out all relative paths.
here::i_am("vaccine-effectiveness/vaccine-effectiveness.Rmd")

# common-setup will attempt to load all the dependencies including 
# ggrrr and dtrackr.
source(here::here("common-setup.R"))

# This creates a function that can be used to put all the outputs (like) a specific directory
out = outputter(directory=here::here("output/vaccine-effectiveness"))

# don't need this yet.
# sup = data_supplement(out = out)
# options(hide.supplementary.tables=TRUE)

# this will set the latest date of a data file that will be loaded by the data load scripts.
# this forces a particular version fo the files for this script.
options(reproduce.at="2022-07-11")

# ignore these options for the time being they are
# for formatting the flowchart.
options(dtrackr.strata_glue="{.value}")
options(dtrackr.strata_sep=", ")
options(dtrackr.default_message="{.count} admissions")
options(dtrackr.show_zero_exclusions=FALSE)

# this file contains all the functions to load the data and some date manipulation functions
```

## Load data

```{r data}

# Load the data from input directory. 
# This assumes a file will exist that starts with AvonCAPLRTDCentralDa_y1..., AvonCAPLRTDCentralDa_y2... in the input directory. _y1, _y2 must be present.
# it will produce one data set from the multiple files.
data_raw = load_data("DeltaVE")
data_raw %>% save_data_source_info(out)
data_norm = data_raw %>% normalise_data()

data_aug = data_norm %>% augment_data() 
v = data_aug %>% get_value_sets()

data_aug = data_aug %>% 
  mutate(
    admission.study_week = study_week(admission.date),
    admin.case_or_control = case_when(
      admission.category == v$admission.category$`Confirmed SARS-CoV-2` ~ "case",
      TRUE ~ "control"
    ) %>% factor(levels = c("case","control"))
  )


# this constructs a helper list which tells you the dataframe columns and factor levels as a list

# typing `v$` at the console should give you some auto-complete support to help navigate the copd_study dataframe.

```

## Study specific pipeline

```{r}
# Might want to change these which are defaulted to the start and end of the data:
start_week = min(data_aug$admission.study_week,na.rm=TRUE)
end_week = max(data_aug$admission.study_week,na.rm=TRUE)
# maybe for example:
# start_week = study_week(as.Date("2021-01-01"))
# end_week = study_week(as.Date("2022-01-01"))

# This sets up the data pipeline
# The data frame is being tracked using dtrackr (https://terminological.github.io/dtrackr/index.html)
# It also does some standard exclusions. 
# Depending on what you want to do these exclusions may need to be reviewed.
ve_study = data_aug %>% 
  comment(.messages = "{.count} patients") %>%
  include_any(
    admission.study_week >= start_week & admission.study_week < end_week ~ "{.count} patients admitted \nbetween {start_date_of_week(start_week)} \n and {start_date_of_week(end_week)}"
  ) %>%
  exclude_all(
      is.na(cohort) ~ "{.excluded} not eligible",
      cohort == "control" & !is.na(genomic.variant) ~ "{.excluded} non covid cases with sequencing result"
  ) %>% 
  group_by(cohort) %>%
  comment("{.count} patients") %>%
  status(
    delta = p_count_if(na.omit(genomic.variant_inferred == v$genomic.variant_inferred$Delta)),
    om = p_count_if(na.omit(genomic.variant_inferred == v$genomic.variant_inferred$Omicron)),
    unk = p_count_if(is.na(genomic.variant_inferred)),
    .headline = "{.strata}",
    .messages = c("{delta} delta cases", "{om} omicron cases", "{unk} other cases")
  ) %>%
  # status(
  #   pneumonias = sum(diagnosis.pneumonia == v$diagnosis),
  #   lrtis = sum(diagnosis.LRTI == v$diagnosis.LRTI$yes),
  #   crdes = sum(diagnosis.exacerbation_of_chronic_respiratory_disease == v$diagnosis.exacerbation_of_chronic_respiratory_disease$yes),
  #   hfs = sum(diagnosis.heart_failure == v$diagnosis.heart_failure$yes),
  #   others = sum(
  #     diagnosis.pneumonia != v$diagnosis.pneumonia$yes & 
  #       diagnosis.LRTI != v$diagnosis.LRTI$yes & 
  #       diagnosis.exacerbation_of_chronic_respiratory_disease != v$diagnosis.exacerbation_of_chronic_respiratory_disease$yes & 
  #       diagnosis.heart_failure != v$diagnosis.heart_failure$yes),
  #   .headline = "{.strata}",
  #   .messages = c(
  #     "Total {.count} admissions:",
  #     "{pneumonias} with Pneumonia",
  #     "{lrtis} with NP-LRTI",
  #     "{crdes} with CRDE",
  #     "{hfs} with HF",
  #     "{others} with other presentations"
  #   )
  # ) %>%
  status(
    unvacc = p_count_if(vaccination.vaccination == v$vaccination.vaccination$Unvaccinated),
    d1 = p_count_if(vaccination.vaccination == v$vaccination.vaccination$`1 dose`),
    d2 = p_count_if(vaccination.vaccination == v$vaccination.vaccination$`2 doses`),
    d3 = p_count_if(vaccination.vaccination == v$vaccination.vaccination$`3 doses`),
    .headline = "{.strata}",
    .messages = c(
      "Total {.count} admissions:",
      "{unvacc} unvaccinated",
      "{d1} with single dose at admission",
      "{d2} with 2 doses at admission",
      "{d3} with 3 doses at admission"
    )
  ) %>%
  exclude_all(
    vaccination.vaccination == v$vaccination.vaccination$`1 dose` ~ "{.excluded} with only one dose at admission"
  ) %>%
  comment("{.count} patients")
# N.B. more data filtering usually expected here. 
# probably want to include_any with COPD 


# This will give a xlsx of all the cases that were excluded
ve_study %>% excluded() %>% openxlsx::write.xlsx(file = out("exclusions.xlsx"))

# This will give you a figure 1 consort flow chart. This is a dtrackr function.
ve_study %>% flowchart(filename = out("fig1-flowchart"))

# Untracking the data frame means that normal dplyr / tidyverse functions will be used rather than 
# the overloaded versions from dtrackr. This will be slightly more reliable, but from this point onwards no changes will be tracked.
ve_data = ve_study %>% untrack() %>% ungroup()

```


```{r}

tmp = ve_data %>% 
  linelist_to_ts(admission.date,
    case_novacc ~ cohort == v$cohort$case & vaccination.vaccination == v$vaccination.vaccination$Unvaccinated,
    control_novacc ~ cohort == v$cohort$control & vaccination.vaccination == v$vaccination.vaccination$Unvaccinated,
    delta_novacc ~ genomic.variant_inferred == v$genomic.variant_inferred$Delta & vaccination.vaccination == v$vaccination.vaccination$Unvaccinated,
    omicron_novacc ~ genomic.variant_inferred == v$genomic.variant_inferred$Omicron & vaccination.vaccination == v$vaccination.vaccination$Unvaccinated,
    
    case_2vacc ~ cohort == v$cohort$case & vaccination.vaccination == v$vaccination.vaccination$`2 doses`,
    control_2vacc ~ cohort == v$cohort$control & vaccination.vaccination == v$vaccination.vaccination$`2 doses`,
    delta_2vacc ~ genomic.variant_inferred == v$genomic.variant_inferred$Delta & vaccination.vaccination == v$vaccination.vaccination$`2 doses`,
    omicron_2vacc ~ genomic.variant_inferred == v$genomic.variant_inferred$Omicron & vaccination.vaccination == v$vaccination.vaccination$`3 doses`,
    
    case_3vacc ~ cohort == v$cohort$case & vaccination.vaccination == v$vaccination.vaccination$`3 doses`,
    control_3vacc ~ cohort == v$cohort$control & vaccination.vaccination == v$vaccination.vaccination$`3 doses`,
    delta_3vacc ~ genomic.variant_inferred == v$genomic.variant_inferred$Delta & vaccination.vaccination == v$vaccination.vaccination$`3 doses`,
    omicron_3vacc ~ genomic.variant_inferred == v$genomic.variant_inferred$Omicron & vaccination.vaccination == v$vaccination.vaccination$`3 doses`
    
  ) %>%
  mutate(across(.cols=-admission.date,.fns = function(x) slider::slide_dbl(x, sum, .before=40, .after = 40),.names = "roll_{col}")) 

```


```{r}

# ggrrr has a set of functions making ggplots behave themselves.
# a prejudiced theme for publication and sensible size defaults are enforced by the following
ggrrr::gg_pedantic()


```

## Plotting stuff

```{r}

caseVsControl = 
  bind_rows(
    tmp %>% mutate(odds_ratio_ve(vaccinatedCase = roll_case_2vacc,unvaccinatedCase = roll_case_novacc,vaccinatedControl = roll_control_2vacc, unvaccinatedControl = roll_control_novacc)) %>% mutate(doses = "two doses"),
    tmp %>% mutate(odds_ratio_ve(vaccinatedCase = roll_case_3vacc,unvaccinatedCase = roll_case_novacc,vaccinatedControl = roll_control_3vacc, unvaccinatedControl = roll_control_novacc)) %>% mutate(doses = "three doses")
  )


```

```{r}
omicronVsControl = 
  bind_rows(
    tmp %>% 
      mutate(odds_ratio_ve(vaccinatedCase = roll_omicron_2vacc,unvaccinatedCase = roll_omicron_novacc,vaccinatedControl = roll_control_2vacc, unvaccinatedControl = roll_control_novacc)) %>% mutate(doses = "two doses"),
    tmp %>% mutate(odds_ratio_ve(vaccinatedCase = roll_omicron_3vacc,unvaccinatedCase = roll_omicron_novacc,vaccinatedControl = roll_control_3vacc, unvaccinatedControl = roll_control_novacc)) %>% mutate(doses = "three doses")
  )
```

```{r}
deltaVsControl = 
  bind_rows(
    tmp %>% mutate(odds_ratio_ve(vaccinatedCase = roll_delta_2vacc,unvaccinatedCase = roll_delta_novacc,vaccinatedControl = roll_control_2vacc, unvaccinatedControl = roll_control_novacc)) %>% mutate(doses = "two doses"),
    tmp %>% mutate(odds_ratio_ve(vaccinatedCase = roll_delta_3vacc,unvaccinatedCase = roll_delta_novacc,vaccinatedControl = roll_control_3vacc, unvaccinatedControl = roll_control_novacc)) %>% mutate(doses = "three doses")
  )
```

```{r}
plotData = bind_rows(
  caseVsControl %>% mutate(comparison = "all cases vs controls"),
  deltaVsControl %>% mutate(comparison = "delta vs controls"),
  omicronVsControl %>% mutate(comparison = "omicron vs controls")
)
  
p = ggplot(plotData, aes(x=admission.date, y=VE*100, ymin=VE.OR.q.0.025*100, ymax=VE.OR.q.0.975*100))+geom_line(aes(colour=doses))+geom_ribbon(aes(fill=doses),alpha=0.2)+coord_cartesian(ylim=c(-50,100))+facet_wrap(vars(comparison),ncol=1)+ylab("VE (%)")
p %>% ggrrr::gg_save_as(out("fig2-ve-case-control"),size = ggrrr::std_size$full)
```


```
